### ISSUE TITLE
RabbitMQ message routing

### ISSUE THREAD

**Issue by quantum5** [DMOJ/site#376]
The current implementation of #183 relies on DMOJ/judge#80 to reject unwanted submissions. Should there be no judges to handle the submission, the same submission will be rejected, requeued, and redelivered, only to be rejected again and again.

To resolve this issue, we must use a `headers` exchange, so that messages are routed based on available language and problem. However, RabbitMQ exchanges rejects all unroutable messages.

The only solution that seems practical is periodically dispatch all QU messages again. However, this suffers from a pretty big flaw in itself, as there is the potential for a race condition.

There is also an issue of backwards compatibility, such massive change without any trace of compatibility will result in significant downtime during implementation. A few hours of downtime (warning: this is the expected time. In reality, it may be days.) for a bug that most people probably won't even notice is not practical.

The suggested solution, hence, is using a `fanout` exchange to dispatch messages to the old submission queue, and to a headers exchange, which would dispatch to new clients. This `headers` exchange is allowed to send messages to `/dev/null`. There will be an additional function added to `amqpd` that receives submissions from the old submission queue, and schedule them to be processed a minute later. If the submission status is still `QU`, schedule it for rejudging. This would solve the judging race while also maintaining compatibility, as old judges will still receive submissions normally, perhaps with a minute delay, but it would still work.

However, this solution suffers the ABA problem. To prevent ABA, we will add a `last_judge` `datetime` field, serving as the judge attempt identification. Hence, we simply filter those submissions that are still queued after a minute, resulting in a maximum two minute wait before a rejudge is attempted.

Now we shall look back to the `fanout` solution, and realize that ultimately, the old queue is not necessary when `MySQL` can store that state, though the old queue can be used to minimize the wait time to almost always one minute. Hence, the `fanout` is ultimately unnecessary. (It is worth noting that backwards compatibility is an excuse to make that option look more attractive, but since there is a better solution, why bother?)

Despite this, attempts to rejudge old submissions may still cause chaos. To resolve this, we shall endeavour to only rejudge the submissions that could actually be rejudged, given the judges currently running, and fail other attempts. Since the admin already counts the amount of successes and failures in rejudging, this is an acceptable solution. This will result in an extra query per rejudge, but since the rejudge operation already runs multiple queries, it's not much worse.

In conclusion, we shall use a `headers` exchange and a cron to resolve this horrible issue with repetitive rejection.

/cc @Xyene 

P.S. It is interesting to note that attempting to write an essay is a valid method of coming up with a solution, as an acceptable alternative to an intelligent conversation.

**Comment by quantum5** [comment]
Clearly message routing does not seem to work. The message would be routed to all possible queues. This means our little scheme is foiled. However, we can have queues created for each and every problem, matched with language, in the format: `dmoj.<problem code>.<language code>`. I cry.

However, all is not lost! According to [this random guy](https://www.quora.com/Can-rabbitMQ-or-zeroMQ-handle-1mil-queues), a queue is rather cheap. We can have every judge declare a queue for every problem and language combination. We cannot use an autodelete queue because the message would be lost if the judge doesn't handle it properly.

On dispatch, we simply declare the queue, which has no effect if the queue already exists. Then we send the message as we did before.

On the receiving end, a channel would be bound to all queues that it can handle. There is a [sketchiness](http://stackoverflow.com/questions/24510310/consume-multiple-queues-in-python-pika) regarding prepopulated queues, and hence we need to switch to async. [Async example](http://pika.readthedocs.org/en/0.10.0/examples/asynchronous_consumer_example.html). This should not be too big of a hassle, as there is an `add_timeout` method, whose functionality we even emulate for pings.

The only concern I have is dead problems with not so dead queues. We can use a daily cron to clean up all queues that have no consumers and no messages.

/cc @Xyene

**Comment by quantum5** [comment]
Hmm, turns out to be a BAD idea. You can't subscribe to all those queues in a reasonable amount of time. Since the multiple binding thing suffers from the same problem, we are back to square zero.

If there is a solution, it probably will be that each judge gets a queue, and we somehow ensure only one judge actually judges it, then have a dead letter queue which we periodically try to refill into newly connected judges.

**Comment by quantum5** [comment]
I think I have a partial solution. Every judge gets its own queue. The database has enough state to allow information to be dispatched to all available judges. Each judge would have another queue, bound to a fanout exchange, that allows every judge to be notified when a submission is claimed. If the judge was to first to claim, it starts judge. However, there would be no redispatch logic, should the judging fail.

**Comment by WallE256** [comment]
I have a question, are we limited at using RabbitMQ, or we can add a new piece of software to the stack?

**Comment by quantum5** [comment]
I doubt this matters anymore, as we decided to remove RabbitMQ from submission dispatching.
